### PR TITLE
[NFC] Make a more explicit SR and Matrix type hierarchy  + add SRJacobian (#648)

### PR DESCRIPTION
**Pull request by PhilipVinc**
This PR makes no change to our public API or functional changes within netket.
It simply creates a class hierarchy for SMatrix types as follows:

```
AbstractSMatrix < -- AbstractLazySMatrix <-- OnTheFlySMatrix <--IterativeSMatrix
```

My idea is that #648 can be rebased on top of this one, and it's Matrix type can be a subclass of `AbstractLazySMatrix`. 
This will be handy also for discussions in #649.

The PR also clearly defines what is the api that those objects should define to conform to NetKet.

### CODE REVIEW

File: netket/optimizer/sr/base.py [hunk 1]
```diff
@@ -32,3 +34,96 @@ class SR:
...
+        any method available, usually but not necessarily defined inside the
+        sr object stored inside the AbstractSMatrix.
+        This function must accept arbitrary additional arguments.
+
+
+    Additionally, you get for free:
+    - :code:`__call__(y)`, meaning you must be able to do S(vec) = S@vec.
+        This is defined by the base class so you don't need to define it.
+        This guarantees that you can pass this matrix to sparse solvers.
+    - :code:`__rtruediv__(y)`, that allows you to solve Sx=y by doing y\\S.
```
Comment (femtobit) [inline netket/optimizer/sr/base.py:69]: Having `x = y \\ S` solve `Sx = y` is a bit confusing: Julia (and matlab) uses `A\b` to solve `Ax = b`, so exactly the other way round. I think this is more difficult to implement here, so we could just keep `solve`. It's not used so often that a shorthand is really necessary IMHO.

File: netket/optimizer/sr/base.py [hunk 2]
```diff
@@ -32,3 +34,96 @@ class SR:
...
+        raise NotImplementedError()
+
+    # PUBLIC API: Only override if you want to, but there should be no need.
+    def __call__(self, vec):
+        return self @ vec
+
+    # PUBLIC API: Only override if you want to, but there should be no need.
+    def __rtruediv__(self, y):
+        x, _ = self.solve(y)
+        return x
```
Comment (femtobit) [inline netket/optimizer/sr/base.py:117]: See above.

File: netket/optimizer/sr/base.py [hunk 3]
```diff
@@ -32,3 +34,96 @@ class SR:
...
+
+    # PUBLIC API: Only override if you want to, but there should be no need.
+    def __array__(self) -> jnp.ndarray:
+        return self.to_dense()
+
+
+@dispatch
+def SMatrix(sr: SR, vstate: object) -> AbstractSMatrix:
+    """
+    Construct the S matrix given a
```
Comment (femtobit) [inline netket/optimizer/sr/base.py:127]: Given a ...?

File: netket/optimizer/sr/s_onthefly_mat.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    SR system x = ⟨S⟩⁻¹⟨F⟩, where S is a lazy linear operator
+
+    Tolerances are applied according to the formula
+    :code:`norm(residual) <= max(tol*norm(b), atol)`
+    """
+
+    centered: bool = struct.field(pytree_node=False, default=True)
+    """Uses S=⟨ΔÔᶜΔÔ⟩ if True (default), S=⟨ÔᶜΔÔ⟩ otherwise. The two forms are 
+    mathematically equivalent, but might lead to different results due to numerical
+    precision. The non-centered variaant should bee approximately 33% faster.
```
Comment (femtobit) [inline netket/optimizer/sr/s_onthefly_mat.py:34]:
```suggestion
    precision. The non-centered variant should be approximately 33% faster.
```

File: netket/variational/mc_state.py [hunk 5]
```diff
@@ -27,6 +27,8 @@
 from flax import linen as nn
 from flax import serialization
 
+from plum import dispatch
+
```
Comment (femtobit) [inline netket/variational/mc_state.py:31]: This seems to be unused here?
```suggestion
```

**Review by femtobit — COMMENTED** [review]
Just some initial comments.

All in all I'm very much in favor of defining a proper S matrix / SR hierarchy here, but I haven't had the time to look at this PR in detail yet.

File: netket/optimizer/sr/base.py (hunk 1 above)
Comment (PhilipVinc) [reply netket/optimizer/sr/base.py:69]: This is `__rtruediv__` so happens when you call `b/A`. But indeed you are right that this is a bit too much.

File: netket/optimizer/sr/base.py (hunk 3 above)
Comment (PhilipVinc) [reply netket/optimizer/sr/base.py:127]: I should have deleted this.
It's part of other changes that I was experimenting with.

**Comment by gcarleo** [comment]
hi, just to understand the need for this hierarchy better: why can't we just derive the S matrix from ```scipy.sparse.linalg.LinearOperator```(or the jax equivalent) with maybe a few more methods, say something to return O_k if needed? Then custom solvers are just implemented as objects that rely on basic operations such as matrix product products or using the O_k ?

**Comment by gcarleo** [comment]
the advantage is that you can use the standard sparse solver with no modification. if you want to implement a slightly modified sparse solver (say, adding a diagonal shift) it is also easy, I think, within this logic. The on the fly is just a way to compute the matrix vector product, so it also fits within

**Comment by gcarleo** [comment]
actually, the S matrix should really just be a linear operator, not even have methods to return O_k, since those are implementation details (for example, for a quantum state on a quantum computer the O_k are not even well defined).

**Comment by attila-i-szabo** [comment]
> hi, just to understand the need for this hierarchy better: why can't we just derive the S matrix from `scipy.sparse.linalg.LinearOperator`(or the jax equivalent) with maybe a few more methods, say something to return O_k if needed? Then custom solvers are just implemented as objects that rely on basic operations such as matrix product products or using the O_k ?

I think  `AbstractSMatrix` should exist as a base class because we should provide a `solve` method that can be overridden by the subclasses with any subclass-specific optimisation. Whether we want to make that a subclass of `LinearOperator` and which one (if not both) is a relatively minor point, the solvers will work with it happily either way as long as `__matmul__` and `__call__` are implemented.

> It simply creates a class hierarchy for SMatrix types as follows:
> AbstractSMatrix < -- AbstractLazySMatrix <-- OnTheFlySMatrix <--IterativeSMatrix

I'm not sure I see the need for so many steps. Could you explain the difference between `IterativeSMatrix` and `OnTheFlySMatrix`? The names would seem more logical the other way around (I can see my `JacobianSMatrix` subclassing something called `IterativeSMatrix` but not the other), but then what is the difference between a lazy and an iterative S matrix?

**Comment by gcarleo** [comment]
> I think AbstractSMatrix should exist as a base class because we should provide a solve method that can be overridden by the subclasses with any subclass-specific optimisation. 

Yes it's just that it is not clear to me that we need the ```solve```method in the first place. All the solvers I can think of are written in terms of matrix vector products only or, in the worst case, converting the linear operator to a full matrix, which is also already supported here. Is there an example where having the matrix vector method is not enough to solve the problem?

**Comment by attila-i-szabo** [comment]
> > I think AbstractSMatrix should exist as a base class because we should provide a solve method that can be overridden by the subclasses with any subclass-specific optimisation.
> 
> Yes it's just that it is not clear to me that we need the `solve`method in the first place. All the solvers I can think of are written in terms of matrix vector products only or, in the worst case, converting the linear operator to a full matrix, which is also already supported here. Is there an example where having the matrix vector method is not enough to solve the problem?

In principle it's enough of course. In practice one would worry about efficiency and numerical stability, so one might want to (for example) precondition the solver with a preconditioner that depends on the gradients, so it would be generated by the S matrix object too. (This can actually be important if one uses a heterogeneous network with very different gradients in different parts, been there, done that.) One could then call the solver from `Vmc` like `gmres(S, f, M=S.preconditioner()), but this requires that all S matrix objects return a preconditioner that might not make sense for some implementations, and is just generally less transparent, because we take care of the functionality of the S matrix somewhere completely different. A `solve` method can deal with all of this as it's best for a given implementation.

**Comment by PhilipVinc** [comment]
> why can't we just derive the S matrix from scipy.sparse.linalg.LinearOperator(or the jax equivalent)

There is no jax equivalent, and in general an S matrix does not really satisfy the LinearOperator interface, as we do not necessarily have a well defined dtype, nor shape (the shape we could declare it to be ^2 the number of parameters, but not the dtype, in general).

However as Attila pointed out, `AbstractSMatrix` (could change the name if that's unclear) requires subclasses to define at least `matrix-vector` product, and provides everything needed to be compatible with sparse solvers in general.
In the future we might decide to subclass it `LinearOperator` but I'd avoid it for now. In any case it will be a one line change if needed.

> actually, the S matrix should really just be a linear operator, not even have methods to return O_k

I agree. That is what this definition requires.

> Yes it's just that it is not clear to me that we need the solvemethod in the first place

It's for performance reasons raised by @attila-i-szabo . We already have it, and this PR is trying not to touch this, but in the future we can add it or remove it. That's what I was trying to discuss in my last comment in #649. Can we keep the discussion there?
In short, the reason for `.solve` to exist is that some S matrix implementations might work better in the `PyTree` basis or in the `dense-ravelled` basis, even if they work with both. 

For example: 
 - the current `OnTheFly` implementation works in the `PyTree` basis, and if you multiply it by a dense vector it will convert it to a PyTree (unravel), perform a matmul then ravel the result. 
 - Attila's `JacobianSMatrix`, which works in the `dense` basis. If you multiply it by a PyTree vector it will first ravel it, perform the matmul, then unravel the result. This means that to use it in a solver you could, in principle, just use it as-is, but you'll get better performance if you ravel the PyTree before calling solve, and unravel the result of solve, rather then ravel-matmul-unravel at every solver iteration.

**Comment by PhilipVinc** [comment]
I would like to repeat that this PR is simply to introduce a class hierarchy and not to change the current API.
It also adds a few utility methods that where missing.

The current Api has oddities, like the solve method, but those are best discussed in #649 .

What might be improved here are the class names.

My idea for the class are

- `AbstractSMatrix -> root class for s matrices defining the interface`
- `AbstractLazySMatrix` -> Abstract base class for lazy s matrices. By lazy s matrices I mean matrices that rely on AD to compute the dense matrix, so they need to have inside the `apply_fun`, `parameters` and `samples` of the model. This could also be called `AbstractJaxSMatrix` as it's really jax-only stuff. But then this could be confusing. `AbstractJaxADSMatrix`?
- `OnTheFlySMatrix` -> On The fly S matrix. It's the implementation by inaluig without specifying anything about the solver used.
-  `SROnTheFlyIterative` -> On the fly S Matrix + the `x0` initial state field. This is for working with iterative solvers.

@attila-i-szabo , you should create a `JacobianSMatrix(AbsstractSMatrix)`. Your implementation is not lazy in the sense of using AD to compute the Ok during the matmul itself. 

`AbstractLazySMatrix` is there because I am playing with another lazy representation and I found it handy.

`OnTheFlySMatrix` and `SROnTheFlyIterative` are almost the same thing. 
The problem is that right now we are conflating the SMatrix representation and the solver objects into the same structures, leading to those oddities...
I suspect if we go for a design like what I propose in #649 they'll converge to the same thing, but for now I want to separate them because if you want to use the OnTheFly with a non-iterative solver you want to subclass `OnTheFlySMatrix`

**Comment by femtobit** [comment]
I think the role of the `MCState.quantum_geometric_tensor(self, sr)` function is strange in this context as it's only purpose is to call `sr.create(self)`. This results in code like this
https://github.com/netket/netket/blob/ab04a5357cb8ceae949c9716c837a2022b2013dc/Test/Optimizer/test_sr_itersolve.py#L110-L116
which is exactly equivalent to
```python
srobj = SRObj(mode="R2R" if vstate.model.dtype is float else "holomorphic")
S = srobj.create(vstate)
x, _ = S.solve(vstate.parameters)
```
So, really, why have a `quantum_geometric_tensor` method if it requires the user to do all the setup of the SR object anyways? It's not like other subclasses of `AbstractVariationalState` could meaningfully do something different than calling `sr.create` in the current design. In that case, we should probably just use `srobj.create(vstate)` [1] directly and remove the implicit circular dependency between `MCState` and the S matrix hierarchy through `quantum_geometric_tensor`.

[1] Or even just pass the `MCState` to the constructor of `SR...`, while we're simplifying things. Then, we could also directly look at the state to provide a meaningful default for `mode` instead of doing it externally like in the test above.

**Comment by attila-i-szabo** [comment]
> [1] Or even just pass the `MCState` to the constructor of `SR...`, while we're simplifying things. 

I think https://github.com/netket/netket/issues/649#issuecomment-829267314 is heading in this direction. It drops `SR*` classes altogether (other than minimal sugar for providing sensible defaults) and replaces `srobj.create` with a call to the S matrix object constructor (with some caveats in the comments below it).

> Then, we could also directly look at the state to provide a meaningful default for `mode` instead of doing it externally like in the test above.

As explained in https://github.com/netket/netket/pull/652#issuecomment-826651844 and https://github.com/netket/netket/pull/652#issuecomment-826739700, choosing a meaningful default is not something easily done. We could expand the Jacobian-based code to allow for it (see https://github.com/netket/netket/pull/652#issuecomment-827949805), but that'd be a nontrivial amount of work. Before that, allowing autodetection of `mode` would be very costly and/or very error-prone, neither of which is something we should aim for.

**Comment by femtobit** [comment]
Ah, yes, thanks @attila-i-szabo; I missed that comment in this slightly split discussion. That indeed seems to address my point very well.

**Comment by PhilipVinc** [comment]
@femtobit indeed you raise a valid point.
In the current (and future) design `vstate.quantum_geom_ten(sr)` is basically useless.

It's there because that's how we designed the variational state interface in the first place.
I think it's ok to keep it even if it does very little, for consistency with `expect` and `expect_and_grad`. But I'm not 100% convinced yet

**Comment by PhilipVinc** [comment]
Closed in favour of #674 going directly to master
